tsl/src/chunk_copy.c

```diff
@@ -625,6 +811,197 @@ chunk_copy(Oid chunk_relid, const char *src_node, const char *dst_node, bool del
+						1,
+						chunk_copy_activity_tuple_found,
+						AccessShareLock,
+						CHUNK_COPY_ACTIVITY_TABLE_NAME,
+						&cc);
+
+	/*
+	 * If a valid entry is returned then fill up the rest of the fields in the
+	 * ChunkCopy structure
+	 */
+	if (cc)
+	{
+		cc->mcxt = mcxt;
+		cc->chunk = ts_chunk_get_by_id(cc->fd.chunk_id, true);
+		cc->stage = NULL;
+
+		/* Was the chunk compressed in the meanwhile? */
+		if (ts_chunk_is_compressed(cc->chunk))
+			ereport(ERROR,
+					(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+					 errmsg("\"%s\" is a compressed remote chunk. Chunk copy/move cleanup not "
+							"supported"
+							" currently on compressed chunks",
+							get_rel_name(cc->chunk->table_id))));
+
+		/* No other sanity checks need to be performed since they were done earlier */
+
+		/* Setup the src_node */
+		cc->src_server =
+			data_node_get_foreign_server(NameStr(cc->fd.source_node_name), ACL_USAGE, true, false);
+		Assert(NULL != cc->src_server);
+
+		/* Setup the dst_node */
+		cc->dst_server =
+			data_node_get_foreign_server(NameStr(cc->fd.dest_node_name), ACL_USAGE, true, false);
+		Assert(NULL != cc->dst_server);
+	}
+
+	MemoryContextSwitchTo(old);
+
+	if (cc == NULL)
+		/* No entry found, long lived context not required */
+		MemoryContextDelete(mcxt);
+
+	return cc;
+}
+
+static void
+chunk_copy_cleanup_internal(ChunkCopy *cc, int stage_idx)
+{
+	bool first = true;
+
+	/* Cleanup each copy stage in a separate transaction */
+	do
+	{
+		StartTransactionCommand();
+
+		cc->stage = &chunk_copy_stages[stage_idx];
+		if (cc->stage->function_cleanup)
+			cc->stage->function_cleanup(cc);
+
+		/* Mark stage as cleaned up and update the catalog */
+		if (!first && stage_idx != 0)
+			chunk_copy_activity_update(cc);
+		else
+			first = false;
+
+		CommitTransactionCommand();
+	} while (--stage_idx >= 0);
+}
+
+void
+chunk_copy_cleanup(int32 activity_id, const char *operation_id)
+{
+	ChunkCopy *cc;
+	const MemoryContext oldcontext = CurrentMemoryContext;
+	const ChunkCopyStage *stage;
+	bool found = false;
+	int stage_idx = -1;
+
+	if (!superuser())
+		ereport(ERROR,
+				(errcode(ERRCODE_INSUFFICIENT_PRIVILEGE),
+				 (errmsg("must be superuser to copy/move chunk to data node"))));
+
+	if (dist_util_membership() != DIST_MEMBER_ACCESS_NODE)
+		ereport(ERROR,
+				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+				 errmsg("function must be run on the access node only")));
+
+	cc = chunk_copy_activity_get(activity_id, operation_id);
+
+	if (cc == NULL)
+		ereport(ERROR,
+				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+				 errmsg("invalid chunk copy activity identifier. Entry not found")));
+
+	/* Identify the last completed stage for this activity. */
+	for (stage = &chunk_copy_stages[0]; stage->name != NULL; stage++)
+	{
+		stage_idx++;
+		if (namestrcmp(&cc->fd.completed_stage, stage->name) == 0)
+		{
+			found = true;
+			break;
+		}
+	}
+
+	/* should always find an entry, add ereport to quell compiler warning */
+	Assert(found == true);
```

Comment:
This can be triggered manually by updating the complete_stage in the table, right?

Reply:
Yeah, adding a bogus stage will trigger this indeed.